### PR TITLE
Add satellite features and add support to the user can direct which security groups are added to their workers

### PR DESCRIPTION
**Pull request by Blintmester**
### Community Note

* Please vote on this pull request by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original pull request comment to help the community and maintainers prioritize this request
* Please do not leave "+1" or other comments that do not add relevant new information or questions, they generate extra noise for pull request followers and do not help prioritize the request

Relates OR Closes #0000

Output from acceptance testing:

```
$ make testacc TEST=./ibm/service/satellite TESTARGS='-run=TestAccSatelliteLocation_PodAndServiceSubnet'

--- PASS: TestAccSatelliteLocation_PodAndServiceSubnet (1187.34s)
PASS
ok      command-line-arguments  1187.409s
```

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -237,8 +237,12 @@ replace github.com/dgrijalva/jwt-go v3.2.0+incompatible => github.com/golang-jwt
 
 replace github.com/portworx/sched-ops v0.0.0-20200831185134-3e8010dc7056 => github.com/portworx/sched-ops v0.20.4-openstorage-rc3 // required by rook v1.7
 
+replace github.com/IBM-Cloud/container-services-go-sdk v0.0.0-20231106114255-c50117860a3c => github.com/Blintmester/container-services-go-sdk v0.0.0-20231116135145-fae5e6201d4c // for development only
```
Comment (z0za) [inline go.mod:240]: I believe these can go, right?

Comment (Blintmester) [reply go.mod]: sure, my mistake, fixed it

**Comment by dawall1mn** [comment]
I see the update to allow security groups on vpc create cluster.  It was not obvious this PR also includes support to specify security groups on vpc worker pool create.  Is that included?

File: ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go [hunk 2]
```diff
@@ -322,6 +322,14 @@ func ResourceIBMContainerVpcCluster() *schema.Resource {
 				RequiredWith:     []string{"kms_instance_id", "crk"},
 			},
 
+			"cluster_security_groups": {
```
Comment (hkantare) [inline ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go:325]: can we rename this argument `security_groups` because the resource already indicates its a cluster

File: ibm/service/satellite/data_source_ibm_satellite_location.go [hunk 3]
```diff
@@ -138,6 +138,18 @@ func DataSourceIBMSatelliteLocation() *schema.Resource {
 					},
 				},
 			},
+			"service_subnet": {
+				Type:        schema.TypeString,
```
Comment (hkantare) [inline ibm/service/satellite/data_source_ibm_satellite_location.go:142]: In datasource these should be computed only remove optional and forceNew.
Same for `pod_subnet` also

**Comment by hkantare** [comment]
Fix conflicts
Update the respective docs files for new argumets added
security_groups were added only for vpc_cluster should we added for worker_pool resource also

File: ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go (hunk 2 above)
Comment (Blintmester) [reply ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go]: sure, I renamed it

File: ibm/service/satellite/data_source_ibm_satellite_location.go (hunk 3 above)
Comment (Blintmester) [reply ibm/service/satellite/data_source_ibm_satellite_location.go:142]: fixed it

**Comment by Blintmester** [comment]
> Fix conflicts in go.mod and go.sum files Update the respective docs files for new argumets added security_groups were added only for vpc_cluster should we added for worker_pool resource also

If I can fix everything, go.mod and go.sum files will be fixed.

Currently I try to add security_groups to worker_pool resource, will check in if its done!

**Comment by Blintmester** [comment]
If everything is fine, I'll fix the go.mod conflicts too.

**Comment by Blintmester** [comment]
@hkantare I pushed the docs files too, please can you review it?

File: ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go [hunk 4]
```diff
@@ -322,6 +322,14 @@ func ResourceIBMContainerVpcCluster() *schema.Resource {
 				RequiredWith:     []string{"kms_instance_id", "crk"},
 			},
 
+			"security_groups": {
```
Comment (z0za) [inline ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go:325]: it looks like this parameter can only be set at create
you can use `DiffSuppressFunc: flex.ApplyOnce` to ignore any changes to the parameter

or

it can be marked with forcenew to recreate the cluster if this param is changed. this means the original cluster will be deleted, so this should be handled with extra care

File: ibm/service/kubernetes/resource_ibm_container_worker_pool.go [hunk 5]
```diff
@@ -42,7 +44,7 @@ func ResourceIBMContainerWorkerPool() *schema.Resource {
 					"cluster"),
 			},
 
-			"machine_type": {
+			"flavor": {
```
Comment (z0za) [inline ibm/service/kubernetes/resource_ibm_container_worker_pool.go:47]: please don't convert this resource to vpc, there is already a vpc_worker_pool one

File: ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go [hunk 6]
```diff
@@ -322,6 +322,14 @@ func ResourceIBMContainerVpcCluster() *schema.Resource {
...
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Description: "Allow user to set which security groups added to their workers",
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Set:         flex.ResourceIBMVPCHash,
+			},
+
```
Comment (z0za) [inline ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go:332]: it looks like it is missing from the API, but...
should we be able to set SecurityGroupIDs to the default workerpool too?

File: website/docs/r/container_worker_pool.html.markdown [hunk 7]
```diff
@@ -52,6 +52,44 @@ resource "ibm_container_worker_pool" "test_pool" {
...
+  security_groups = [ 
+  "r134-ee951766-31e7-4fdb-bde8-0f08315b0cc6", # one
+  "r134-dab9930e-cf2d-46d5-9808-9e52955f15f2", # two
+  "r134-4f29f2fb-979d-451b-bac7-1e6c773e63d7", # three
+  #"cluster",
+  ]
+
+  zones {
+      name = "us-south-1"
+      subnet_id = "0716-2389b003-7968-41b1-833c-4d63d7165899"
```
Comment (z0za) [inline website/docs/r/container_worker_pool.html.markdown:73]: please remove the real ids from the example. you can add variables instead

File: ibm/service/satellite/resource_ibm_satellite_location.go [hunk 8]
```diff
@@ -209,6 +209,20 @@ func ResourceIBMSatelliteLocation() *schema.Resource {
 				Computed:  true,
 				Sensitive: true,
 			},
+			"service_subnet": {
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+				Description: "Custom subnet CIDR to provide private IP addresses for services",
+				//Computed:    true,
```
Comment (z0za) [inline ibm/service/satellite/resource_ibm_satellite_location.go:217]: if it is not needed, please remove it

File: ibm/flex/structures.go [hunk 9]
```diff
@@ -3242,13 +3263,13 @@ func FlattenOpaqueSecret(fields containerv2.Fields) []map[string]interface{} {
 	return flattenedOpaqueSecret
 }
 
-// flattenHostLabels ..
-func FlattenHostLabels(hostLabels []interface{}) map[string]string {
+// flatten the provided key-value pairs
+func FlattenKeyValues(keyValues []interface{}) map[string]string {
```
Comment (z0za) [inline ibm/flex/structures.go:3267]: why did you rename this function?

File: ibm/service/satellite/resource_ibm_satellite_cluster.go [hunk 10]
```diff
@@ -280,6 +280,12 @@ func ResourceIBMSatelliteCluster() *schema.Resource {
 				Sensitive:   true,
 				Description: "The IBM Cloud Identity and Access Management (IAM) service CRN token for the service that creates the cluster.",
 			},
+			"calico_ip_autodetection": {
```
Comment (z0za) [inline ibm/service/satellite/resource_ibm_satellite_cluster.go:283]: is it only for create too? maybe ApplyOnce  or forcenew?

**Review by z0za — CHANGES_REQUESTED** [review]

File: ibm/flex/structures.go (hunk 9 above)
Comment (Blintmester) [reply ibm/flex/structures.go:3267]: previously it was used only for flatten Host Labels, but I renamed it and we can use it for everything when we should flatten key-values (map[string]string), and I used it too (and it was not host labels)

File: ibm/service/satellite/resource_ibm_satellite_cluster.go (hunk 10 above)
Comment (Blintmester) [reply ibm/service/satellite/resource_ibm_satellite_cluster.go:283]: yes, it is only for create, so I'll add ApplyOnce

File: ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go (hunk 4 above)
Comment (lewiseevans) [reply ibm/service/kubernetes/resource_ibm_container_vpc_cluster.go:325]: I think we should go with ApplyOnce rather than ForceNew.

At the moment, these values can not be changed in the API. If the ability to change the security groups was added in the future and customers update the security groups outside of TF, TF would replace the cluster.

If we go with ApplyOnce, the changes would be ignored until the TF provider supported that functionality at which point we could provide guidance to customers on importing the updated changes and avoid cluster deletion.

File: ibm/service/kubernetes/resource_ibm_container_vpc_worker_pool.go [hunk 11]
```diff
@@ -212,6 +212,14 @@ func ResourceIBMContainerVpcWorkerPool() *schema.Resource {
 				Computed:    true,
 				Description: "Autoscaling is enabled on the workerpool",
 			},
+
+			"security_groups": {
```
Comment (z0za) [inline ibm/service/kubernetes/resource_ibm_container_vpc_worker_pool.go:216]: can you please add the ApplyOnce diff suppress here too?

File: ibm/service/kubernetes/resource_ibm_container_vpc_cluster_test.go [hunk 12]
```diff
@@ -152,6 +152,34 @@ func TestAccIBMContainerVpcClusterDedicatedHost(t *testing.T) {
...
 }
 
+func TestAccIBMContainerVpcClusterSecurityGroups(t *testing.T) {
+	name := fmt.Sprintf("tf-vpc-cluster-%d", acctest.RandIntRange(10, 100))
+	var conf *v2.ClusterInfo
+
+	resource.Test(t, resource.TestCase{
+		PreCheck:  func() { acc.TestAccPreCheck(t) },
+		Providers: acc.TestAccProviders,
+		// CheckDestroy: testAccCheckIBMContainerVpcClusterDestroy,
```
Comment (z0za) [inline ibm/service/kubernetes/resource_ibm_container_vpc_cluster_test.go:162]: why is it commented out?

File: ibm/service/kubernetes/resource_ibm_container_vpc_worker_pool.go (hunk 11 above)
Comment (z0za) [inline ibm/service/kubernetes/resource_ibm_container_vpc_worker_pool.go:216]: can you please add a test for this too?

File: website/docs/r/satellite_location.html.markdown [hunk 13]
```diff
@@ -44,6 +44,23 @@ resource "ibm_satellite_location" "create_location" {
...
+data "ibm_resource_group" "group" {
+    name = "Default"
+}
+
+resource "ibm_satellite_location" "create_location" {
+  location          = var.location
+  zones             = var.location_zones
+  managed_from      = var.managed_from
+  resource_group_id = data.ibm_resource_group.group.id
+  pod_subnet = var.pod_subnet // "10.42.0.0/16"
```
Comment (z0za) [inline website/docs/r/satellite_location.html.markdown:59]: can you please fix the formatting in the example?

File: ibm/service/kubernetes/resource_ibm_container_vpc_worker_pool.go (hunk 11 above)
Comment (Blintmester) [reply ibm/service/kubernetes/resource_ibm_container_vpc_worker_pool.go:216]: sure, will do

File: ibm/service/kubernetes/resource_ibm_container_vpc_cluster_test.go (hunk 12 above)
Comment (Blintmester) [reply ibm/service/kubernetes/resource_ibm_container_vpc_cluster_test.go:162]: the CheckDestroy function fails on every run, because it doesn't wait enough, or something else, I don't really know, but the resource (after a failed run) was destroyed successfully

**Review by z0za — APPROVED** [review]

**Review by lewiseevans — APPROVED** [review]

**Review by hkantare — APPROVED** [review]
